### PR TITLE
Add NoUnneededFinalMethodFixer

### PR DESCRIPTION
**Pull request by Slamdunk**
```diff
 final class Foo {
-    final public function foo() {}
-    final protected function bar() {}
-    final private function baz() {}
+    public function foo() {}
+    protected function bar() {}
+    private function baz() {}
 }
```

### CODE REVIEW

**Comment by keradus** [comment]
Not sure about the rule itself.
Same way, one could say that `public` in `public function` is unneeded

File: src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            ]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        return $tokens->isAnyTokenKindsFound([T_CLASS]) && $tokens->isAnyTokenKindsFound([T_FINAL]);
```
Comment (julienfalque) [inline src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: You can replace these `isAnyTokenKindsFound()` calls with a single `isAllTokenKindsFound()` class.

**Comment by Slamdunk** [comment]
> Not sure about the rule itself.
> Same way, one could say that `public` in `public` function is unneeded

@keradus you are right; my reasoning is:

1) I want a *standard*, be it with or without final; having final classes with half final methods and half non-final methods is nonsense

2) What standard should we pursue?

Well, visibility needs to be explicit because in this example:
```php
class A {
    protected function foo() {}
}
class B extends A {
    function foo() {}
}
```
PHP has strictly rules about the result, but the developer could be confused by what should be the outcome (hit: `B::foo()` becomes `public`)

On the other side, `final` keyword in a `final` class has no interaction with possible parent classes and as far as I can see there is no misunderstanding of what the outcome is, that is the same with or without final.

So my choise is to raise verbosity for `visibility`, and to reduce verbosity for `final`

File: src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php (hunk 1 above)
Comment (Slamdunk) [inline src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: Done, thank you

**Comment by keradus** [comment]
why raising verbosity in one place and reducing in another? this looks not consistent

**Comment by Slamdunk** [comment]
Because as I said in one context it helps, in the other it doesn't.

BTW I can make it configurable if you prefer.

**Comment by keradus** [comment]
I agree that some enforced standard to have about this is nice.
While final methods in final class could look superfluous, they follow defensive approach:
when having final class + final methods, if one drop final from class, then he need to decide which methods he want to open as well. comparing to final only on class level - droping one final would open every method.

**Comment by Slamdunk** [comment]
This is a good point.

In our business this can't happen because we adopt two harsh custom fixers:
1. [All non-abstract classes MUST be final](https://github.com/Slamdunk/php-cs-fixer-extensions/blob/v1.4.0/lib/FinalInternalClassFixer.php)
1. [All abstract classes MUST have public methods marked ad final](https://github.com/Slamdunk/php-cs-fixer-extensions/blob/v1.4.0/lib/FinalAbstractPublicFixer.php)

To reduce inheritance nightmares and increase template method pattern.

I told you this not advertise something, but to drop some thoughts about the topic.
We would never benefit from a fixer that marks methods as final in a final class.

**Comment by keradus** [comment]
FYI: #1389 ;)

If you go extra defensive like that, it's cool and not needed for you indeed.
But when one don't follow that approach and has only some of the class marked as final, he would benefit from it.

**Comment by Slamdunk** [comment]
Knew #1389 👍 but never used because it requires an annotation to be written; I use PHP-CS-Fixer in CI to get rid of lazy developers 😸 

It seems we all understood the topic, but what about the solution?
Is it ok for you to make it configurable, with default behaviour as adding the `final` keyword?
If yes, @julienfalque please help me to find a better fixer name, you words master 🙏

**Comment by keradus** [comment]
To be honest, not sure... yet ;)
I do see the value of the fixer, I find your reasoning valid in your case for sure.
On the other side, I want to promote only good, generic rules. That's the reason why we don't have option to remove `public` before `function`.
Not claiming this rule in current form is bad, don't take it like that. I just need to think a bit more about it ;)

Other opinions are welcome of course ;)

**Comment by julienfalque** [comment]
IMO this PR can be merged as is. I don't see the benefits of a defensive approach with `final` keyword on each method. I would not make the fixer configurable, and its name looks good to me :)

**Comment by keradus** [comment]
btw, if you would add it to `.php_cs.dist`, what will happen ?

File: src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            if (!$tokens[$index]->isGivenKind(T_CLASS)) {
+                continue;
+            }
+
+            $classOpen = $tokens->getNextTokenOfKind($index, ['{']);
+            $classClose = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $classOpen);
+
+            $prevToken = $tokens[$tokens->getPrevMeaningfulToken($index)];
+            if ($prevToken->isGivenKind(T_FINAL)) {
+                $this->fixClass($tokens, $classOpen, $classClose);
```
Comment (keradus) [inline src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: you basically traverse whole tokens of single class twice - one to find it's end, one to fix it's content.
not needed, you may determine it's end while fixing it.
pass only `$classOpen` here,
then, in `fixClass`, if you see nested `{`, it's some method body opening, you skip whole block.
if you find `}`, it's close `}` of class. voila, you just found a `$classClose`, no need to pass it to `fixClass` method, no need to determine it in `applyFix`, and you could return it via `fixClass` to you could still do the jump `$index = $classClose`

**Comment by Slamdunk** [comment]
> btw, if you would add it to `.php_cs.dist`, what will happen ?

Nothing because the only final method here is this one, and is legitime:

https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/63661f3add3609e90e4ab8115113e189ae547bb4/src/AbstractFixer.php#L66

**Comment by keradus** [comment]
please add it to `.php_cs.dist` file

**Comment by Slamdunk** [comment]
Already done

**Comment by keradus** [comment]
then, my other concert is still in place ;)

**Comment by Slamdunk** [comment]
I'm working 🏃

File: src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php (hunk 2 above)
Comment (Slamdunk) [inline src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: New logic applied.

FYI, this code was copy-pasted from [ProtectedToPrivateFixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.4.0/src/Fixer/ClassNotation/ProtectedToPrivateFixer.php), and the algorithm is a @SpacePossum optimization: https://github.com/Slamdunk/PHP-CS-Fixer/pull/1

After this, we should refactor other fixers too.

File: src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+     * @param Tokens $tokens
+     * @param int    $classOpenIndex
+     * @param int    $end
+     * @param bool   $isFinalClass
+     *
+     * @return int
+     */
+    private function fixClass(Tokens $tokens, $classOpenIndex, $end, $isFinalClass)
+    {
+        for ($index = $classOpenIndex + 1; $index < $end; ++$index) {
```
Comment (keradus) [inline src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: `end` is never reached.
then, when you have `open` for class, `end` for file is weird ;)

File: src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * @param Tokens $tokens
+     * @param int    $classOpenIndex
+     * @param int    $end
+     * @param bool   $isFinalClass
+     *
+     * @return int
+     */
+    private function fixClass(Tokens $tokens, $classOpenIndex, $end, $isFinalClass)
```
Comment (keradus) [inline src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: if class is final we could just jump over whole class, right? no need for iterating it manuallty, jumping over single methods and so

File: src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php (hunk 3 above)
Comment (Slamdunk) [inline src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: Yes `end` is never reached, but without it we have to use a `while` and I dislike loops without a declared end, it is far easier to end in an infinite loop.
I'll change the name in the next commit.

File: src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php (hunk 4 above)
Comment (Slamdunk) [inline src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: Right. I just tried to write less code, but it seems is less readable.

File: src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php (hunk 3 above)
Comment (keradus) [inline src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: if `$end === $tokens->count()`, no need to pass duplicated data

Comment (Slamdunk) [inline src/Fixer/ClassNotation/NoUnneededFinalMethodFixer.php]: Done

File: tests/Fixer/ClassNotation/NoUnneededFinalMethodFixerTest.php [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            ],
+            'preserve-comment' => [
+                '<?php final class Foo { /* comment */public function foo() {} }',
+                '<?php final class Foo { final/* comment */public function foo() {} }',
+            ],
+            'multiple-classes-per-file' => [
+                '<?php final class Foo { public function foo() {} } abstract class Bar { final public function bar() {} }',
+                '<?php final class Foo { final public function foo() {} } abstract class Bar { final public function bar() {} }',
+            ],
+
```
Comment (julienfalque) [inline tests/Fixer/ClassNotation/NoUnneededFinalMethodFixerTest.php]: Is this empty line intended?

Comment (Slamdunk) [inline tests/Fixer/ClassNotation/NoUnneededFinalMethodFixerTest.php]: I find it usseful to separate to-be-changed cases from dont-touch ones

**Comment by SpacePossum** [comment]
@keradus can you look at the failing utest? https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer/jobs/266908788#L757 (not sure who to ping)

This PR looks good to me :)
I would however like to see a utest to be added with multiple candidates in one test/file (like `<?php class A {} class B{}`). Pretty sure the fixer will work correctly, but the added test will prevent regression.

**Comment by keradus** [comment]
you already approved a fix for that failure #2989

**Comment by julienfalque** [comment]
@SpacePossum This failure is unrelated to changes here and will be fixed by https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2989.

**Comment by keradus** [comment]
Thank you @Slamdunk.

**Pull request closed without merging**